### PR TITLE
set upstream after pulling a PR

### PR DESCRIPTION
**Pull request by randy3k**
It is useful to set the upstream to the repo of the PR author because github lets repo admin push to the PR branch.

### CODE REVIEW

File: github/commands/pull_request.py [hunk 1]
```diff
@@ -79,13 +79,13 @@ def on_select_action(self, idx):
         elif idx == 1:
             show_single_line_input_panel(
                 "Enter branch name for PR {}:".format(self.pr["number"]),
-                "{}/{}".format(self.pr["user"]["login"], self.pr["head"]["ref"]),
```
Comment (randy3k) [inline github/commands/pull_request.py:82]: also change the default value of branch name from owner/repo to owner-repo to avoid ambiguity.

File: github/commands/pull_request.py [hunk 2]
```diff
@@ -101,17 +101,20 @@ def fetch_and_checkout_pr(self, branch_name=None):
...
         )
 
-        if branch_name:
-            self.window.status_message("Creating local branch for PR...")
-            self.git(
-                "branch",
-                branch_name,
-                self.pr["head"]["sha"]
-            )
+        if not branch_name:
```
Comment (kaste) [inline github/commands/pull_request.py:104]: This change is not ok since it breaks the "Checkout detached" feature. I mean `branch_name` can be `None` intentionally.

File: github/commands/pull_request.py [hunk 3]
```diff
@@ -101,17 +101,20 @@ def fetch_and_checkout_pr(self, branch_name=None):
...
+            "branch",
+            branch_name,
+            self.pr["head"]["sha"]
+        )
 
         self.window.status_message("Checking out PR...")
-        self.checkout_ref(branch_name or self.pr["head"]["sha"])
+        self.checkout_ref(branch_name)
         util.view.refresh_gitsavvy_interfaces(self.window, refresh_sidebar=True)
+        sublime.set_timeout(lambda: self.set_upstream_for_branch(branch_name))
```
Comment (kaste) [inline github/commands/pull_request.py:117]: Avoid `set_timeout` here. Why is that needed. Also call `refresh_gitsavvy` at the end.

File: github/commands/pull_request.py [hunk 4]
```diff
@@ -121,12 +124,48 @@ def create_branch_for_pr(self, branch_name):
             self.pr["head"]["ref"]
         )
 
+        if not branch_name:
+            branch_name = self.pr["head"]["sha"]
+
```
Comment (kaste) [inline github/commands/pull_request.py:129]: We did not catch `not branch_name` before but here in that case the user did not enter a branch name, and we should probably return/abort the operation.

File: github/commands/pull_request.py [hunk 5]
```diff
@@ -121,12 +124,48 @@ def create_branch_for_pr(self, branch_name):
...
         self.window.status_message("Creating local branch for PR...")
         self.git(
             "branch",
             branch_name,
             self.pr["head"]["sha"]
         )
+        sublime.set_timeout(lambda: self.set_upstream_for_branch(branch_name))
```
Comment (kaste) [inline github/commands/pull_request.py:136]: Avoid `set_timeout`.

File: github/commands/pull_request.py [hunk 6]
```diff
@@ -121,12 +124,48 @@ def create_branch_for_pr(self, branch_name):
...
+                elif index == 0:
+                    url = clone_url
+                elif index == 1:
+                    url = ssh_url
+
+                self.git("remote", "add", remote, url)
+                self.git("fetch", remote, remote_branch)
+                self.git("branch", "-u", "{}/{}".format(remote, remote_branch), branch_name)
+
+            sublime.set_timeout(
```
Comment (kaste) [inline github/commands/pull_request.py:162]: Avoid `set_timeout`

File: github/commands/pull_request.py [hunk 7]
```diff
@@ -121,12 +124,48 @@ def create_branch_for_pr(self, branch_name):
...
+                lambda: self.window.show_quick_panel(
+                    [clone_url, ssh_url], on_select_url)
```
Comment (kaste) [inline github/commands/pull_request.py:164]: I don't know what `clone_url` vs. `ssh_url` actually represent. I will check that on the functional tests later.

Comment (kaste) [reply github/commands/pull_request.py:164]: That being said, Github recommends `https` urls these days because these are accessible and just work in most environments.

File: github/commands/pull_request.py [hunk 8]
```diff
@@ -121,12 +124,48 @@ def create_branch_for_pr(self, branch_name):
...
+            )
+        else:
+            self.git("fetch", remote, remote_branch)
```
Comment (kaste) [inline github/commands/pull_request.py:167]: Here and above L159: we actually already fetched at the beginning of this action. 

```
        self.git(
            "fetch",
            self.pr["head"]["repo"]["clone_url"],
            self.pr["head"]["ref"]
        )
```

Why is that needed? Is there any difference? Not if the user selects `clone_url`, if the user selects `ssh` it just uses a different transport protocol, the ref is the same, and we already have it. No?

File: github/commands/pull_request.py (hunk 7 above)
Comment (randy3k) [reply github/commands/pull_request.py:164]: `clone_url` is like "https://github.com/octocat/Hello-World.git" and `ssh_url` is like "git@github.com:octocat/Hello-World.git"

File: github/commands/pull_request.py (hunk 8 above)
Comment (randy3k) [reply github/commands/pull_request.py:167]: it is needed for "Checkout as detached HEAD." where we do need to set upstream. Perhaps we should prompt user to choose what protocol to fetch.

Comment (randy3k) [reply github/commands/pull_request.py]: Now we always prompt user to choose a protocol when checkout as detached head.

Comment (kaste) [reply github/commands/pull_request.py]: Why is the protocol a matter here? It never was before. As I said these days using https is just fine. Is there an issue open requesting and explaining the need? If this is a real issue we should deduce: if the user already used only non https for the remotes, we do it here as well. Otherwise we use https.

It confuses me to set an upstream if the user chooses detached head. That doesn't make any sense to me. We could ask to add the remote, but iirc the user can just checkout the same PR as a branch later if she needs to add commits. I mean, you only need the remote and upstream if you want to edit the PR, and that doesn't play well with a detached HEAD.

File: github/commands/pull_request.py [hunk 9]
```diff
@@ -94,39 +94,84 @@ def on_select_action(self, idx):
...
+
+        remotes = list(self.get_remotes().keys())
+        remote = self.pr["user"]["login"]
+        remote_branch = self.pr["head"]["ref"]
+
+        clone_url = self.pr["head"]["repo"]["clone_url"]
+        ssh_url = self.pr["head"]["repo"]["ssh_url"]
+
+        if remote not in remotes:
+            if not sublime.ok_cancel_dialog("Add remote '{}'?".format(remote)):
```
Comment (kaste) [inline github/commands/pull_request.py:144]: Now you cannot checkout a PR without adding the remote. Is this intended?

File: github/commands/pull_request.py (hunk 8 above)
Comment (kaste) [reply github/commands/pull_request.py]: Ah you already changed that and don't set the upstream in the detached case. So the only question is why the protocol matters now.

Comment (randy3k) [reply github/commands/pull_request.py]: In Linux, I think it is more common/easier to use ssh key rather than `credential.helper`. 

Also, the https protocol is particularly painful if two-factor authentication is used.

File: github/commands/pull_request.py (hunk 9 above)
Comment (randy3k) [reply github/commands/pull_request.py:144]: I don't know if it is a good default. Do you think we should proceed without setting upstream?

Comment (randy3k) [reply github/commands/pull_request.py:144]: Or we modify the behavior, `Checkout as local branch` sets upstream and `Create local branch` does not.

File: github/commands/pull_request.py (hunk 8 above)
Comment (asfaltboy) [reply github/commands/pull_request.py]: I feel users should be allowed to choose, regardless of github recommendation. Also, it's important that we don't break any behaviour for existing users

Comment (kaste) [reply github/commands/pull_request.py]: The user already made that choice. We can take whatever is set up before. Just look at the already configured remote urls and go with https if it's in there. I don't see why I should make this decision here over and over again. If you made the initial clone using https, just stick with it.

Comment (randy3k) [reply github/commands/pull_request.py]: If the remote exists, you won't be prompted to choose a URL. However, user will be asked if they are fetching the PR as detached head or the remote doesn't exist.

Comment (kaste) [reply github/commands/pull_request.py]: The remote of the main repo determines which protocol I want to use and have set up. Since we're checking out a fork here there must be at least the main repo already configured.

Comment (randy3k) [reply github/commands/pull_request.py]: Perhaps the easier solution is to make a setting `pull_request_protocol` with default value `"ask"` and possible values "ssh" and "https".

**Comment by randy3k** [comment]
By the way, we also need to port the code to `GsGitlabMergeRequestCommand`. Or even better, make the code more CI agnostic.

File: github/commands/pull_request.py (hunk 8 above)
Comment (kaste) [reply github/commands/pull_request.py]: No that's the most complicated solution. Doesn't `self.get_integrated_remote_url()` already gives us the current base remote url, we query for this at the beginning of the process, and so we just look if `base_remote.starts_with("https:")`.

Comment (randy3k) [reply github/commands/pull_request.py]: Alright. It makes sense. I believe that most people don’t need to switch between protocol. I have to go sleeping now, will do the changes tomorrow.

**Comment by kaste** [comment]
Note that the flake8 errors are fixed in dev now, you should drop this commit here.

File: github/commands/pull_request.py [hunk 10]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
+        owner = self.pr["head"]["repo"]["owner"]["login"]
+
+        if owner == self.base_remote.owner:
+            # don't add prefix for integrated remote
+            branch_name = self.pr["head"]["ref"]
+        else:
+            branch_name = "{}-{}".format(owner, self.pr["head"]["ref"])
+
         if idx == 0:
             self.fetch_and_checkout_pr()
```
Comment (kaste) [inline github/commands/pull_request.py:88]:
```suggestion
            self.checkout_detached()
```

File: github/commands/pull_request.py [hunk 11]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
         elif idx == 1:
             show_single_line_input_panel(
                 "Enter branch name for PR {}:".format(self.pr["number"]),
-                "{}/{}".format(self.pr["user"]["login"], self.pr["head"]["ref"]),
+                branch_name,
                 self.fetch_and_checkout_pr
```
Comment (kaste) [inline github/commands/pull_request.py:93]:
```suggestion
                lambda branch_name: self.create_branch_for_pr(branch_name, checkout=True)
```

File: github/commands/pull_request.py [hunk 12]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
             )
         elif idx == 2:
             show_single_line_input_panel(
                 "Enter branch name for PR {}:".format(self.pr["number"]),
-                "{}/{}".format(self.pr["user"]["login"], self.pr["head"]["ref"]),
-                self.create_branch_for_pr
+                branch_name,
+                lambda x: self.create_branch_for_pr(x, ask_set_upstream=True)
```
Comment (kaste) [inline github/commands/pull_request.py:99]:
```suggestion
                lambda branch_name: self.create_branch_for_pr(branch_name, checkout=True)
```

File: github/commands/pull_request.py [hunk 13]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
         util.view.refresh_gitsavvy_interfaces(self.window, refresh_sidebar=True)
 
-    def create_branch_for_pr(self, branch_name):
-        self.window.status_message("Fetching PR commit...")
-        self.git(
-            "fetch",
-            self.pr["head"]["repo"]["clone_url"],
-            self.pr["head"]["ref"]
-        )
+    def create_branch_for_pr(self, branch_name, checkout=False, ask_set_upstream=False):
```
Comment (kaste) [inline github/commands/pull_request.py:130]:
```suggestion
    def create_branch_for_pr(self, branch_name, checkout=False, ask_set_upstream=True):
```

File: github/commands/pull_request.py [hunk 14]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
         if branch_name:
-            self.window.status_message("Creating local branch for PR...")
-            self.git(
-                "branch",
-                branch_name,
-                self.pr["head"]["sha"]
-            )
+            self.create_branch_for_pr(branch_name, checkout=True, ask_set_upstream=True)
+        else:  # detached HEAD
+            self.checkout_detached()
```
Comment (kaste) [inline github/commands/pull_request.py:110]: Remove method; it also has a bug (or is it a feature?) that if the user did not enter a branch name we continue with `checkout_detached` instead of aborting. If it's a feature we should have it of course for case 1 *and* 2 t.i. on L132.

File: github/commands/pull_request.py [hunk 15]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
+        clone_url = self.pr["head"]["repo"]["clone_url"]
+        ssh_url = self.pr["head"]["repo"]["ssh_url"]
+        ref = self.pr["head"]["ref"]
+
+        if self.base_remote_url.startswith("https://"):
+            url = clone_url
+        elif self.base_remote_url.startswith("git@"):
+            url = ssh_url
+        else:
+            raise RuntimeError("Cannot recognize url {}.".format(self.base_remote_url))
```
Comment (kaste) [inline github/commands/pull_request.py:124]: Simplify and be open, e.g. 

```
url = ssh_url if self.base_remote_url.startswith("git@") else clone_url
```

File: github/commands/pull_request.py [hunk 16]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
+                "Set upstream to '{}/{}'?".format(owner, ref))
+
+        if set_upstream:
+            if owner not in remotes.keys():
+                if self.base_remote_url.startswith("https://"):
+                    url = clone_url
+                elif self.base_remote_url.startswith("git@"):
+                    url = ssh_url
+                else:
+                    raise RuntimeError("Cannot recognize url {}.".format(self.base_remote_url))
```
Comment (kaste) [inline github/commands/pull_request.py:160]: Same as above, usually be open

```
url = ssh_url if self.base_remote_url.startswith("git@") else clone_url
```

File: github/commands/pull_request.py [hunk 17]
```diff
@@ -30,8 +30,10 @@ def run(self):
         sublime.set_timeout_async(self.run_async, 0)
 
     def run_async(self):
-        base_remote = github.parse_remote(self.get_integrated_remote_url())
-        self.pull_requests = github.get_pull_requests(base_remote)
+        self.base_remote_name = self.get_integrated_remote_name()
```
Comment (kaste) [inline github/commands/pull_request.py:33]: `self.get_integrated_remote_name()` is costly as it does at min 2 git calls, but the result is only used in a conditional once. Note that `get_integrated_remote_url` calls this internally.
```suggestion
```

File: github/commands/pull_request.py [hunk 18]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
-            "branch",
-            branch_name,
-            self.pr["head"]["sha"]
-        )
+        set_upstream = False
+
+        remotes = self.get_remotes()
+        owner = self.pr["head"]["repo"]["owner"]["login"]
+        if owner == self.base_remote.owner:
+            owner = self.base_remote_name
```
Comment (kaste) [inline github/commands/pull_request.py:140]:
```suggestion
            owner = self.get_integrated_remote_name()
```

Can you explain what this does. First, owner is `self.pr["head"]["repo"]["owner"]["login"]` which is probably a GitHub username? Here we change the `owner` to `base_remote_name`, which is the local remote name(?), so we have for example `owner == origin` here?

File: github/commands/pull_request.py [hunk 19]
```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
...
+        self.git("branch", branch_name, ref)
+        self.git("branch", "-u", ref, branch_name)
+
+        if checkout:
+            self.checkout_ref(branch_name)
+
+        util.view.refresh_gitsavvy_interfaces(self.window, refresh_sidebar=True)
+
+    def create_branch_from_sha_for_pr(self, branch_name, checkout):
+        self.git("branch", branch_name, self.pr["head"]["sha"])
```
Comment (kaste) [inline github/commands/pull_request.py:180]: I think you forgot the "fetch" here.

Generally I don't understand why `create_branch_from_remote_for_pr` and `create_branch_from_sha_for_pr` are here, and especially why they differ so much.  From the code flow you would say they both should do the same, except one variant *also* sets upstream which roughly are one or two git calls. 

So they should be the same code, except:

```
self.git("remote", "add", owner, url)  # if the remote is new
self.git("branch", "-u", ref, branch_name)  # set actually new side-effect, setting the upstream
```

Can't we do 

```
self.git("fetch", url, self.pr["head"]["ref"])
self.git("branch", branch_name, self.pr["head"]["sha"])
if set_upstream:
	self.git("remote", "add", remote, url)
	self.git("branch", "-u", "{}/{}".format(remote, remote_branch), branch_name)
if checkout:
	self.checkout_ref(branch_name)
```

File: github/commands/pull_request.py (hunk 18 above)
Comment (randy3k) [reply github/commands/pull_request.py:140]: `self.pr["head"]["repo"]["owner"]["login"]` could be an github user/organization name. For example, it could be `timbrel` if a PR is coming from a branch of `GitSavvy`. It is checking if the `base_remote` owner is the same as the PR's owner. 

In most cases, `self.base_remote.owner` would be the username of the PR submitter and `self.base_remote_name` would be `"origin".`

File: github/commands/pull_request.py (hunk 19 above)
Comment (randy3k) [reply github/commands/pull_request.py:180]: In order to set up stream
```
self.git("branch", "-u", "{}/{}".format(remote, remote_branch), branch_name)
```

We need to make sure that the remote branch is fetched from the remote.

```
        self.git("fetch", remote, remote_branch)
        ref = "{}/{}".format(remote, remote_branch)
        self.git("branch", branch_name, ref)
        self.git("branch", "-u", ref, branch_name)
```
and it cannot be simply
```
self.git("fetch", url, self.pr["head"]["ref"])
``

File: github/commands/pull_request.py [hunk 20]
```diff
@@ -74,59 +77,95 @@ def on_select_action(self, idx):
...
+        self.git("branch", branch_name, ref)
+        self.git("branch", "-u", ref, branch_name)
+
+        if checkout:
+            self.checkout_ref(branch_name)
+
+        util.view.refresh_gitsavvy_interfaces(self.window, refresh_sidebar=True)
+
+    def create_branch_from_sha_for_pr(self, branch_name, remote, checkout):
+        self.git("fetch", remote, self.pr["head"]["sha"])
```
Comment (kaste) [inline github/commands/pull_request.py:162]: Honestly, I feel stupid, but for this branch of the code, t.i. we don't set upstream, we can't assume we have a remote set. Can we? We need to fetch from the URL I would have thought.

Also why Sha and not ref?

File: github/commands/pull_request.py (hunk 19 above)
Comment (kaste) [reply github/commands/pull_request.py:163]: So there is a difference between git fetch URL ref and git fetch remote which points to URL ref.

Comment (randy3k) [reply github/commands/pull_request.py:163]: Yes. 

```
self.git("fetch", url, self.pr["head"]["ref"])
```
would fetch the ref to `FETCH_HEAD` instead.

File: github/commands/pull_request.py (hunk 20 above)
Comment (randy3k) [reply github/commands/pull_request.py:162]: Oh, you are right.

File: github/commands/pull_request.py (hunk 19 above)
Comment (kaste) [reply github/commands/pull_request.py]: Haha, after 53 comments you introduce a new concept "FETCH_HEAD". 

> would fetch the ref to FETCH_HEAD instead

Yeah, okay, but then `git branch branch_name FETCH_HEAD`. I argue that you only need the `remote/remote_branch` for the `branch -u` call. It may not matter, I would like the least moving parts/variants as possible for readability. 

Otherwise, I think it's okay now. 

BTW. Have you seen "View diff" is broken. Github changed its server configuration and now sends lowercase headers.  Basically we look for "Location: ..." and they send "location: ...". Or "content-type"...

Comment (randy3k) [reply github/commands/pull_request.py]: > I argue that you only need the remote/remote_branch for the branch -u call

I am not sure if I understand.

> BTW. Have you seen "View diff" is broken. 

Maybe, I never used this option.

**Comment by kaste** [comment]
Can you take a look on the refactoring.  Now a typical "checkout as branch" looks like this:

```
 (D) [ 94ms] $ git remote -v
 (D) [ 78ms] $ git config --local --get GitSavvy.ghRemote
 (M) [784ms] $ git fetch https://github.com/timbrel/GitSavvy.git review
 (M) [ 78ms] $ git branch review2 FETCH_HEAD
 (M) [ 94ms] $ git checkout review2
 (M) [ 78ms] $ git branch -u origin/review review2
```

I don't see why this shouldn't work. Is this stupid? Note that I reordered here, we fetch and checkout immediately, and ask to set the upstream after. T.i. we don't ask if the previous commands failed.

The second commit is to remove one further `get_remotes()` call.

**Comment by randy3k** [comment]
We will need to fetch from remote explicitly after `remote add`.

**Comment by kaste** [comment]
Oh, fuck, you're right 🙄  but really I only understand and see this now you wrote it and every git command is aligned, t.i. after the refactoring. Simple code line by line.

**Comment by randy3k** [comment]
LGTM. Thank for reviewing it.

**Comment by kaste** [comment]
🔥

**Comment by kaste** [comment]
Love the feature
